hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/IncrementalRelationV2.scala

```diff
@@ -207,78 +207,83 @@ class IncrementalRelationV2(val sqlContext: SQLContext,
       val sOpts = optParams.filter(p => !p._1.equalsIgnoreCase("path"))
 
       val startInstantArchived = !queryContext.getArchivedInstants.isEmpty
-      val endInstantTime = queryContext.getEndInstant.get()
-
-      val scanDf = if (fallbackToFullTableScan && startInstantArchived) {
-        log.info(s"Falling back to full table scan as startInstantArchived: $startInstantArchived")
-        fullTableScanDataFrame(commitsToReturn)
+      if (queryContext.isEmpty) {
+        // no commits to read
+        sqlContext.sparkContext.emptyRDD[Row]
       } else {
-        if (filteredRegularFullPaths.isEmpty && filteredMetaBootstrapFullPaths.isEmpty) {
-          sqlContext.createDataFrame(sqlContext.sparkContext.emptyRDD[Row], usedSchema)
-        } else {
-          log.info("Additional Filters to be applied to incremental source are :" + filters.mkString("Array(", ", ", ")"))
+        val endInstantTime = queryContext.getLastInstant
 
-          var df: DataFrame = sqlContext.createDataFrame(sqlContext.sparkContext.emptyRDD[Row], usedSchema)
+        val scanDf = if (fallbackToFullTableScan && startInstantArchived) {
+          log.info(s"Falling back to full table scan as startInstantArchived: $startInstantArchived")
+          fullTableScanDataFrame(commitsToReturn)
+        } else {
+          if (filteredRegularFullPaths.isEmpty && filteredMetaBootstrapFullPaths.isEmpty) {
```

Comment:
Can we keep the change incremental and most of the lines unchanged?

Reply:
> Can we keep the change incremental and most of the lines unchanged?

Done~